connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java

```diff
@@ -185,46 +180,105 @@ public Task newTask(Class<? extends Task> taskClass) {
+        return t.cast(o);
     }
 
-    public Converter newConverter(String converterClassOrAlias, AbstractConfig config) {
-        Class<? extends Converter> klass;
-        try {
-            klass = pluginClass(
-                    delegatingLoader,
-                    converterClassOrAlias,
-                    Converter.class
-            );
-        } catch (ClassNotFoundException e) {
-            throw new ConnectException(
-                    "Failed to find any class that implements Converter and which name matches "
-                            + converterClassOrAlias
-                            + ", available converters are: "
-                            + pluginNames(delegatingLoader.converters())
-            );
+    /**
+     * Instantiate and configure a {@link Converter} identified in the supplied configuration with the given configuration property name.
+     * This method first attempts to find the specified class using the current class loader, and if that's not available it looks for
+     * a plugin with the class.
+     *
+     * @param config            the configuration containing the {@link Converter}'s configuration; may not be null
+     * @param classPropertyName the name of the property that contains the name of the {@link Converter} class; may not be null
+     * @param isKeyConverter    true if this converter is a key converter, or false otherwise
+     * @return the instantiated and configured {@link Converter}; never null
+     * @throws ConnectException if the {@link Converter} implementation class could not be found
+     */
+    public Converter newConverter(AbstractConfig config, String classPropertyName, boolean isKeyConverter) {
+        // First see if we can instantiate the instance using the current classloader ...
+        Converter plugin = getInstance(config, classPropertyName, Converter.class);
+        if (plugin == null) {
+            // Not found on the current classloader, so look in other plugins ...
+            String converterClassOrAlias = config.getClass(classPropertyName).getName();
+            Class<? extends Converter> klass;
+            try {
+                klass = pluginClass(
+                        delegatingLoader,
+                        converterClassOrAlias,
+                        Converter.class
+                );
+            } catch (ClassNotFoundException e) {
+                throw new ConnectException(
+                        "Failed to find any class that implements Converter and which name matches "
+                                + converterClassOrAlias
+                                + ", available converters are: "
+                                + pluginNames(delegatingLoader.converters())
+                );
+            }
+            plugin = newPlugin(klass);
         }
-        return config != null ? newConfiguredPlugin(config, klass) : newPlugin(klass);
+        if (plugin instanceof Configurable) {
```

Comment:
This still doesn't seem correct. We're only invoking configuration if the plugin implements `Configurable` afaict. This does not work for `Converter`s implemented against the new API and assuming the "forward" configuration. We *must* always invoke the "old" `configure(Map, boolean)`, and only invoke the `Configurable` version as a backup. Possibly it would make sense to indicate on the `HeaderConverter` that the `Configurable` methods should be idempotent if we need to be able to implement both.

Not sure if we can test this easily with unit tests, but I think we might want a plain old `Converter` (that does not implement `HeaderConverter`) in tests to validate compatibility... but it's possible we'd need either integration or system tests to properly validate.

Reply:
Another way to put this is that maybe we should make sure our built-in converters can handle calls to both `configure(Map, boolean isKey)` followed by `configure(Map)` with the `TYPE_CONFIG`. then, only things that are `Configurable` see the second one. old implementations wouldn't see it as they would not have implemented `Configurable` (except in unusual circumstances). New implementations could be warned by docs in `HeaderConverer` that they should take care to handle that sequence of calls.